File: src/main/docker.ts
// Docker management, compose-scoped, over the `docker` CLI. We orchestrate the
// existing tool — no Docker Desktop reimplementation. Status is a fact from
// `docker compose ps`; actions shell out to compose. Everything is defensive:
// a missing CLI or a compose error becomes a reported state, never a throw.

import { execFile } from 'node:child_process'
import { existsSync } from 'node:fs'
import { join } from 'node:path'
import { promisify } from 'node:util'
import type { DockerAction, DockerContainer, DockerStatus } from '../shared/types'

const exec = promisify(execFile)
const COMPOSE_FILES = ['docker-compose.yml', 'docker-compose.yaml', 'compose.yml', 'compose.yaml']

function composeFile(root: string): string | null {
  for (const f of COMPOSE_FILES) {
    if (existsSync(join(root, f))) return f
  }
  return null
}

async function run(root: string, args: string[], timeout = 15_000): Promise<string> {
  const { stdout } = await exec('docker', args, { cwd: root, timeout, maxBuffer: 4 * 1024 * 1024 })
  return stdout
}

/** `docker compose ps` emits either JSON-lines or a JSON array depending on version. */
function parsePs(stdout: string): DockerContainer[] {
  const text = stdout.trim()
  if (!text) return []
  const rows: Record<string, unknown>[] = []
  try {
    const asArray = JSON.parse(text)
    if (Array.isArray(asArray)) rows.push(...asArray)
    else rows.push(asArray)
  } catch {
    for (const line of text.split('\n')) {
      const l = line.trim()
      if (!l) continue
      try {
        rows.push(JSON.parse(l))
      } catch {
        /* skip unparseable line */
      }
    }
  }
  return rows.map((r) => ({
    id: String(r.ID ?? r.Id ?? ''),
    name: String(r.Name ?? r.Names ?? ''),
    service: String(r.Service ?? ''),
    state: String(r.State ?? '').toLowerCase(),
    status: String(r.Status ?? r.Health ?? ''),
    ports: String(r.Ports ?? r.Publishers ?? '')
  }))
}

export class DockerService {
  async status(root: string): Promise<DockerStatus> {
    const empty: DockerStatus = { available: false, hasCompose: false, containers: [], error: '' }
    if (!root) return empty

    const file = composeFile(root)
    if (!file) return { ...empty, available: true, hasCompose: false }

    try {
      const out = await run(root, ['compose', 'ps', '--format', 'json', '--all'])
      return { available: true, hasCompose: true, containers: parsePs(out), error: '' }
    } catch (e) {
      const err = e as { code?: string; stderr?: string; message?: string }
      // ENOENT => docker binary not found.
      if (err.code === 'ENOENT') {
        return { available: false, hasCompose: true, containers: [], error: 'docker CLI не найден' }
      }
      return {
        available: true,
        hasCompose: true,
        containers: [],
        error: (err.stderr || err.message || 'ошибка docker').trim().slice(0, 500)
      }
    }
  }

  async action(root: string, action: DockerAction): Promise<{ ok: boolean; error: string }> {
    if (!root || !composeFile(root)) return { ok: false, error: 'нет compose-файла' }
    const args =
      action === 'up'
        ? ['compose', 'up', '-d']
        : action === 'down'
          ? ['compose', 'down']
          : ['compose', 'restart']
    try {
      await run(root, args, 120_000)
      return { ok: true, error: '' }
    } catch (e) {
      const err = e as { stderr?: string; message?: string }
      return { ok: false, error: (err.stderr || err.message || 'ошибка').trim().slice(0, 500) }
    }
  }

  async logs(root: string, service?: string): Promise<string> {
    if (!root || !composeFile(root)) return ''
    const args = ['compose', 'logs', '--tail', '200', '--no-color']
    if (service) args.push(service)
    try {
      return await run(root, args, 20_000)
    } catch (e) {
      const err = e as { stderr?: string; message?: string }
      return (err.stderr || err.message || '').trim()
    }
  }
}

File: src-tauri/src/docker.rs
// Docker management via the `docker` CLI, compose-scoped (mirror of src/main/docker.ts).

use std::path::Path;
use std::process::Command;

use crate::types::{DockerContainer, DockerStatus, OpResult};

const COMPOSE_FILES: &[&str] = &[
    "docker-compose.yml",
    "docker-compose.yaml",
    "compose.yml",
    "compose.yaml",
];

fn has_compose(root: &str) -> bool {
    COMPOSE_FILES.iter().any(|f| Path::new(root).join(f).exists())
}

fn docker(root: &str, args: &[&str]) -> Result<String, (bool, String)> {
    match Command::new("docker").current_dir(root).args(args).output() {
        Ok(out) => {
            if out.status.success() {
                Ok(String::from_utf8_lossy(&out.stdout).to_string())
            } else {
                Err((true, String::from_utf8_lossy(&out.stderr).trim().to_string()))
            }
        }
        Err(e) => {
            // ENOENT-ish → docker not found
            let not_found = e.kind() == std::io::ErrorKind::NotFound;
            Err((!not_found, e.to_string()))
        }
    }
}

fn parse_ps(stdout: &str) -> Vec<DockerContainer> {
    let text = stdout.trim();
    if text.is_empty() {
        return vec![];
    }
    let mut rows: Vec<serde_json::Value> = vec![];
    if let Ok(v) = serde_json::from_str::<serde_json::Value>(text) {
        if let Some(arr) = v.as_array() {
            rows.extend(arr.clone());
        } else {
            rows.push(v);
        }
    } else {
        for line in text.lines() {
            if let Ok(v) = serde_json::from_str::<serde_json::Value>(line.trim()) {
                rows.push(v);
            }
        }
    }
    let get = |r: &serde_json::Value, keys: &[&str]| -> String {
        for k in keys {
            if let Some(s) = r.get(k).and_then(|v| v.as_str()) {
                return s.to_string();
            }
        }
        String::new()
    };
    rows.iter()
        .map(|r| DockerContainer {
            id: get(r, &["ID", "Id"]),
            name: get(r, &["Name", "Names"]),
            service: get(r, &["Service"]),
            state: get(r, &["State"]).to_lowercase(),
            status: get(r, &["Status", "Health"]),
            ports: get(r, &["Ports", "Publishers"]),
        })
        .collect()
}

pub fn status(root: &str) -> DockerStatus {
    let mut s = DockerStatus::default();
    if root.is_empty() {
        return s;
    }
    if !has_compose(root) {
        s.available = true;
        return s;
    }
    match docker(root, &["compose", "ps", "--format", "json", "--all"]) {
        Ok(out) => {
            s.available = true;
            s.has_compose = true;
            s.containers = parse_ps(&out);
        }
        Err((available, msg)) => {
            s.available = available;
            s.has_compose = true;
            s.error = if available {
                msg.chars().take(500).collect()
            } else {
                "docker CLI не найден".to_string()
            };
        }
    }
    s
}

pub fn action(root: &str, action: &str, service: Option<String>) -> OpResult {
    if root.is_empty() || !has_compose(root) {
        return OpResult {
            ok: false,
            error: "нет compose-файла".to_string(),
        };
    }
    let mut args: Vec<String> = vec!["compose".to_string()];
    match action {
        "up" => {
            args.push("up".to_string());
            args.push("-d".to_string());
        }
        "down" => args.push("down".to_string()),
        other => args.push(other.to_string()), // restart | start | stop
    }
    if action != "down" {
        if let Some(svc) = service {
            args.push(svc);
        }
    }
    let refs: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
    match docker(root, &refs) {
        Ok(_) => OpResult {
            ok: true,
            error: String::new(),
        },
        Err((_, msg)) => OpResult {
            ok: false,
            error: msg.chars().take(500).collect(),
        },
    }
}
